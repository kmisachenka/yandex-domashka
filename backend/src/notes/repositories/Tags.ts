import { Tag } from '../../types';

export default class Tags {
  private tags: Tag[];

  private constructor(tags: Tag[]) {
    this.tags = tags;
  }

  public static factory(tags: Tag[]): Tags {
    return new Tags(tags);
  }

  public toArray(): Tag[] {
    return this.tags;
  }

  public addTag(tag: Tag): void {
    // Нет проверки на то что такой тег уже есть
    this.tags.push(tag);
  }

  public hasTag(id: number): boolean {
    const tag = this.tags.find(t => t.id === Number(id));
    return Boolean(tag);
  }
}
