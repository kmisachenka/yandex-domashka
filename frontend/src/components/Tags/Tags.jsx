import React from 'react';
import * as PropTypes from 'prop-types';

import styles from './Tags.module.scss';

export default function Tags(props) {
  const { tags, allTags } = props;
  const labels = allTags.filter(tag => tags.some(id => id === tag.id));
  return (
    <ul className={styles.tags}>
      {labels.map(label => (
        <li key={label.id} className={styles.tag}>
          {label.tag}
        </li>
      ))}
    </ul>
  );
}

Tags.propTypes = {
  tags: PropTypes.arrayOf(PropTypes.number).isRequired,
  allTags: PropTypes.arrayOf(PropTypes.object).isRequired,
};
