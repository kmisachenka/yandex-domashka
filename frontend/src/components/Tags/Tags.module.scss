.tags {
  display: flex;
  flex-wrap: wrap;
  list-style: none;
  margin: 0;
  padding: 0;
  padding-bottom: 16px;
}

.tag {
  background-color: rgba(51, 51, 51, 0.1);
  border-radius: 16px;

  max-width: 210px;

  font-family: Helvetica, serif;
  font-size: 14px;
  line-height: 22px;
  margin: 12px 12px 0 0;
  overflow: hidden;
  padding: 2px 12px;
  text-align: center;

  text-overflow: ellipsis;
  white-space: nowrap;

  transition: 0.2s background-color;

  &:last-child {
    margin-bottom: 0;
    margin-right: 0;
  }

  &:hover {
    background-color: rgba(51, 51, 51, 0.2);
  }
}
