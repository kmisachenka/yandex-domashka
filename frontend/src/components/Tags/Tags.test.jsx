import React from 'react';
import { shallow } from 'enzyme';

import Tags from './Tags';

const allTags = [
  {
    id: 0,
    tag: 'покупки',
  },
  {
    id: 1,
    tag: 'Работа',
  },
  {
    id: 2,
    tag: 'ШРИ',
  },
];

describe('Tags', () => {
  it('should match snapshot', () => {
    const tags = [0, 1, 2];
    const wrapper = shallow(<Tags tags={tags} allTags={allTags} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should render only provided tags', () => {
    const tags = [1];
    const wrapper = shallow(<Tags tags={tags} allTags={allTags} />);
    expect(wrapper).toMatchSnapshot();
  });
});
