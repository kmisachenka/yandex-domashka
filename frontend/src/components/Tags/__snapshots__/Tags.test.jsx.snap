// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tags should match snapshot 1`] = `
<ul
  className="tags"
>
  <li
    className="tag"
    key="0"
  >
    покупки
  </li>
  <li
    className="tag"
    key="1"
  >
    Работа
  </li>
  <li
    className="tag"
    key="2"
  >
    ШРИ
  </li>
</ul>
`;

exports[`Tags should render only provided tags 1`] = `
<ul
  className="tags"
>
  <li
    className="tag"
    key="1"
  >
    Работа
  </li>
</ul>
`;
